README
The Debian Package gnome-exe-thumbnailer
----------------------------

This package was made based on an idea first seen in Ubuntu Brainstorm.  It was
then written as part of the Karmic-Wine-Integration spec for Ubuntu:
https://wiki.ubuntu.com/karmic-wine-integration

It is my hope that this can be merged upstream and become a part of a default
Gnome desktop, as I feel it adds a lot of usability even when you don't have
Wine installed.

 -- Scott Ritchie <scottritchie@ubuntu.com>  Tue, 18 Aug 2009 02:33:41 -0700


CHANGELOG
gnome-exe-thumbnailer (0.9.5) unstable; urgency=high

  * Add a fallback thumbnail_limit value of 10485760 bytes if Nautilus isn't installed
  * Switch to msitools' msiinfo for ProductVersion fetching. This replaces the insecure
    VBScript-based parsing, which has issues described at
    http://news.dieweltistgarnichtso.net/posts/gnome-thumbnailer-msi-fail.html
  * Use a dark background colour for the version label so that it is consistently visible

 -- James Lu <bitflip3@gmail.com>  Tue, 18 Jul 2017 08:08:18 +0800

gnome-exe-thumbnailer (0.9.4-0ubuntu1) unstable; urgency=medium

  * New upstream version
     - respect org.gnome.nautilus.preferences thumbnail-limit (LP: #602767)
     - thumbnail templates for Captiva, Compass, Evolvere, NITRUX-Buttons,
       Paper, Square-Beam, and Vibrancy icon themes

 -- Jan Nekvasil <jan@nekvasil.cz>  Tue, 12 May 2015 15:06:06 +0200

gnome-exe-thumbnailer (0.9.3-0ubuntu1) trusty; urgency=low

  * New upstream version
     - application/x-ms-shortcut (.lnk) support,
       thanks to jopka (https://launchpad.net/~jopka)
     - templates for the Moka icon theme
     - look for the usable icons with lower bit depths 
       when the current one contains only an empty box 
       (e.g. in Simon Tatham's Portable Puzzle Collection)
     - support both ~/.thumbnails and ~/.cache/thumbnails storages
     - erase appropriate old thumbnails in all thumbnails/* 
       subfolders after theme change, not only in thumbnails/normal
  * Update standards to 3.9.5

 -- Scott Ritchie <scottritchie@ubuntu.com>  Wed, 26 Feb 2014 20:29:55 -0800

gnome-exe-thumbnailer (0.9-0ubuntu2) raring; urgency=low

  * Mark package Multi-Arch: foreign (LP: #1123710)

 -- Scott Ritchie <scottritchie@ubuntu.com>  Fri, 15 Feb 2013 17:10:27 -0800

gnome-exe-thumbnailer (0.9-0ubuntu1) precise; urgency=low

  * New version (thanks Jan Nekvasil)
    - added support for GNOME 3 (LP: #752578, #879267)
    - read current icon theme from dconf first
    - added thumbnail templates for the Faience icon theme
    - cleaner templates for the Faenza icon themes
    - DISPLAY=NONE for getting rid of the occasional "Wine is updating" window
      when extracting ProductVersion with MS Windows Script Host
    - delete old thumbnails created by g-e-t on it's first run after the icon
      theme change
  * debian/copyright: convert to dep5 format

 -- Scott Ritchie <scottritchie@ubuntu.com>  Sun, 11 Mar 2012 05:18:23 -0700

gnome-exe-thumbnailer (0.8-0ubuntu1) natty; urgency=low

  * New version (thanks Jan Nekvasil)
    - New style
    - Follows icon themes
  * Commment out code that calls Wine for msi files
  * Simplify debian/rules

 -- Scott Ritchie <scottritchie@ubuntu.com>  Wed, 13 Apr 2011 20:33:19 -0700

gnome-exe-thumbnailer (0.7-0ubuntu1) maverick; urgency=low

  * New version (thanks Jan Nekvasil)
    - support Vista icons (LP: #622220)
  * Include copy of LGPL 2.1 (LP: #648200)
  * Put images in /usr/share/pixmaps/gnome-exe-thumbnailer

 -- Scott Ritchie <scottritchie@ubuntu.com>  Wed, 29 Sep 2010 03:44:35 -0700

gnome-exe-thumbnailer (0.6-0ubuntu1) maverick; urgency=low

  * New version (thanks Jan Nekvasil)
    - use of %i (local file with full path) instead of %u
    - avoid dependency on python
    - only thumbnail local files (LP: #602767)
    - better version number extraction pipeline (LP: #622295)
    - enhanced text positioning with rounded corners in version labels

 -- Scott Ritchie <scottritchie@ubuntu.com>  Fri, 17 Sep 2010 02:04:27 -0700

gnome-exe-thumbnailer (0.5-0ubuntu1) maverick; urgency=low

  * New font with easier readability (LP: #604391)

 -- Scott Ritchie <scottritchie@ubuntu.com>  Sat, 07 Aug 2010 23:49:29 -0700

gnome-exe-thumbnailer (0.5-0ubuntu1~lucidppa1) lucid; urgency=low

  * New font with easier readability (LP: #604391)

 -- Scott Ritchie <scottritchie@ubuntu.com>  Sat, 07 Aug 2010 22:28:19 -0700

gnome-exe-thumbnailer (0.4-0ubuntu1) maverick; urgency=low

  * Port PPA changes to maverick

 -- Scott Ritchie <scottritchie@ubuntu.com>  Sun, 11 Jul 2010 17:10:14 -0700

gnome-exe-thumbnailer (0.4-0ubuntu1~lucidppa2) lucid; urgency=low

  * Reduce font size by one point (still larger than original)

 -- Scott Ritchie <scottritchie@ubuntu.com>  Sat, 10 Jul 2010 17:26:50 -0700

gnome-exe-thumbnailer (0.4-0ubuntu1~lucidppa1) lucid; urgency=low

  * Use fixed-width font and increase size for readability
    - idea thanks to Sven Augustin
  * Use two temporary files to avoid an error message in icoutils

 -- Scott Ritchie <scottritchie@ubuntu.com>  Fri, 02 Jul 2010 01:24:25 -0700

gnome-exe-thumbnailer (0.3-0ubuntu1) maverick; urgency=low

  * Port PPA changes to maverick

 -- Scott Ritchie <scottritchie@ubuntu.com>  Tue, 15 Jun 2010 20:36:09 -0700

gnome-exe-thumbnailer (0.3~lucid1) lucid; urgency=low

  * New template and script that looks nice
  * Thanks to Jan Nekvasil <jan.nekvasil@ryant.cz> for new script and templates

 -- Scott Ritchie <scottritchie@ubuntu.com>  Tue, 15 Jun 2010 19:36:50 -0700

gnome-exe-thumbnailer (0.2) karmic; urgency=low

  * Decode the url-encoded filename passed to wrestool so it actually works
  * Depend on python for the above

 -- Scott Ritchie <scottritchie@ubuntu.com>  Tue, 20 Oct 2009 03:38:25 -0700

gnome-exe-thumbnailer (0.1) karmic; urgency=low

  * Initial Release.  First version only has a template for Wine.

 -- Scott Ritchie <scottritchie@ubuntu.com>  Tue, 18 Aug 2009 02:33:41 -0700


gnome-exe-thumbnailer.sh
#!/bin/bash
shopt -s nocasematch

TEMPFILE1=$(mktemp)
TEMPFILE2=$(mktemp)
TEMPTHUMB=$(mktemp)

INPUTFILE="$1"
OUTPUTFILE="$2"

# Get the current icon theme (or override it by 3rd parameter):
THEME="${3:-$(gconftool-2 --get /desktop/gnome/interface/icon_theme)}"

case "$THEME" in
	Faenza*)
		THEME='faenza'
		DRAW='roundRectangle 2,2 45,45 3,3'
	;;
	
	elementary*|Ubuntu-Mono*|Humanity*)
		THEME='elementary'
		DRAW='roundRectangle 2,2 45,45 3,3'
	;;

	gnome|Human)
		THEME='gnome'
		DRAW='roundRectangle 2,2 45,45 4,4'
	;;

	Breathe)
		THEME='breathe'
		DRAW='roundRectangle 1,2 46,45 2,2'
	;;
	
	Tango*|*)
		THEME='tango'
		DRAW='roundRectangle 2,2 45,45 4,4'
	;;

esac


if [[ ${INPUTFILE##*.} = 'msi' ]]
then
	# Use generic installer icon for a .msi package:
	ICON=/usr/share/pixmaps/gnome-exe-thumbnailer/$THEME/installer.png
	TUNE='-modulate 120,100,0'

else
	# Extract group_icon resource.
	# If we get the "wrestool: $INPUTFILE could not find `1' in `group_icon' resource." error,
	# there is a 99.9% chance that input file is an installer.

	# Warning: Some redirection magic ahead.

	wrestool --extract --type=group_icon "$INPUTFILE" 2>&1 >$TEMPFILE1 | grep 'could not find'

	if [ $? -eq 0 ]
	then
		# Use generic installer icon:
		ICON=/usr/share/pixmaps/gnome-exe-thumbnailer/$THEME/installer.png
		TUNE='-modulate 120'

	else
		# Process extracted data, if we have some:
		if [ -s $TEMPFILE1 ]
		then
			# Look for the best usable icon:
			read OFFSET INDEX < <(
				icotool --list $TEMPFILE1 | awk '{
					ci=int(substr($2,index($2,"=") + 1));
					cw=int(substr($3,index($3,"=") + 1));
					cb=int(substr($5,index($5,"=") + 1));

					if ((cw > w && cw <= 32) || (cw == w && cb > b)) {
						b = cb;
						w = cw;
						i = ci;
					}
				}
				END {
					print (32 - w) / 2, i;
				}'
			)

			# Use a resized 48x48 icon if 32x32 or smaller isn't available.
			# This is very rare, but it happens sometimes:
			if [ "$INDEX" = '' ]
			then
				INDEX=1
				RESIZE=yes
				OFFSET=$(($OFFSET - 12))
			fi

			# Finally try to extract chosen icon:
			icotool --extract --index=$INDEX $TEMPFILE1 -o $TEMPFILE2

			if [ -s $TEMPFILE2 ]
			then
				ICON=$TEMPFILE2
				[ "$RESIZE" ] && mogrify -resize 24x24 $ICON

			else
				# This case generally happens when the hi-res icons are in new "Vista" icon format (bunch of compressed PNGs).
				# Icotool from icoutils 0.29.1 supports it already, but is unable to extract the one selected icon only.

				# Try to extract all icons:
				icotool --extract $TEMPFILE1 -o /tmp

				# There's always a 32x32x32 icon in "Vista" icons, but just to be sure:
				[ -s ${TEMPFILE1}_${INDEX}_32x32x32.png ] && ICON=${TEMPFILE1}_${INDEX}_32x32x32.png

			fi
		fi
	fi
fi


# Create the basic thumbnail:

if [ "$ICON" ]
then
	# Calculate the backgroud color:
	COLOR=$(
		convert $ICON -background white -flatten -fill white \
		-fuzz 40% -opaque black -level 33%,66% -scale 1x1! $TUNE txt:- \
		| tail -1 \
		| grep -o '#......'
	)

else
	# Just use the generic icon with backgroud color based on md5sum:
	HUE=$(md5sum "$INPUTFILE" | cut -c 1-2)
	HUE=$(printf "%d" 0x$HUE)
	COLOR="hsb($HUE, 50%, 90%)"

	LABEL=${INPUTFILE##*/}
	LABEL=$(sed 's/^./\U&/; s/.$/\L&/' <<< "${LABEL:0:2}")

	# Dim color for non-executable files:
	if [[ ! ${INPUTFILE##*.} = 'exe' ]]
	then
		LIGHT=80
		TUNE_NX='-modulate 100,20'
	fi

	convert -size 48x48 xc:none -gravity center -font Helvetica-Narrow-Bold -pointsize 24 \
	-fill '#0000005C' -annotate +1+3 "$LABEL" \
	-fill "hsb($HUE, 3%, ${LIGHT:-100}%)" -annotate +0+2 "$LABEL" \
	png:$TEMPFILE1

	ICON=$TEMPFILE1
	OFFSET=-8

fi

# Create the final thumbnail:
OFFSET=$(($OFFSET + 8))

convert -size 48x48 xc:none -fill "$COLOR" -draw "$DRAW" $TUNE_NX miff:- \
| composite -compose multiply /usr/share/pixmaps/gnome-exe-thumbnailer/$THEME/template.png - png:- \
| composite -geometry +$OFFSET+$OFFSET $ICON - $TEMPTHUMB


# Get the version number:
if [[ ${INPUTFILE##*.} = 'msi' ]]
then
#	# Look for the ProductVersion property if user has the Microsoft (R) Windows Script Host installed:
#	if [ -s "$HOME/.wine/drive_c/windows/system32/cscript.exe" ]
#	then
#		# Workaround wine bug #19799: cscript crashes if you call WScript.Arguments(0)
#		# http://bugs.winehq.org/show_bug.cgi?id=19799
#		<<< "
#			Dim WI, DB, View, Record
#			Set WI = CreateObject(\"WindowsInstaller.Installer\")
#			Set DB = WI.OpenDatabase(\"$INPUTFILE\",0)
#			Set View = DB.OpenView(\"SELECT Value FROM Property WHERE Property = 'ProductVersion'\")
#			View.Execute
#			Wscript.Echo View.Fetch.StringData(1)
#		" iconv -f utf8 -t unicode > $TEMPFILE1.vbs
#
#		VERSION=$(
#			wine cscript.exe //E:vbs //NoLogo Z:\\tmp\\${TEMPFILE1##*/}.vbs 2>/dev/null \
#			| egrep -o '^[0-9]+\.[0-9]+(\.[0-9][0-9]?)?(beta)?'
#		)
#
#	else
		# Try to get the version number from extended file properties at least:
		VERSION=$(
			file "$INPUTFILE" \
			| grep -o ', Subject: .*, Author: ' \
			| egrep -o '[0-9]+\.[0-9]+(\.[0-9][0-9]?)?(beta)?' \
			| head -1
		)
#	fi

else
	# Extract raw version resource:
	wrestool --extract --raw --type=version "$INPUTFILE" > $TEMPFILE1

	if [ -s $TEMPFILE1 ]
	then
		# Search for a sane version string.
		# This (especially the final regexp) took me really long time to figure out. Am I that lame?
		VERSION=$(< $TEMPFILE1 \
			tr '\0, ' '\t.\0' \
			| sed 's/\t\t/_/g' \
			| tr -c -d '[:print:]' \
			| sed -r -n 's/.*Version[^0-9]*([0-9]+\.[0-9]+(\.[0-9][0-9]?)?).*/\1/p'
		)
	fi
fi


# Put a version label on the thumbnail:
if [ "$VERSION" ]
then
	convert -font -*-clean-medium-r-*-*-6-*-*-*-*-*-*-* \
	-background transparent -fill white label:"$VERSION" \
	-trim -bordercolor '#00001090' -border 2 \
	-fill '#00001048' \
	-draw $'color 0,0 point\ncolor 0,8 point' -flop \
	-draw $'color 0,0 point\ncolor 0,8 point' -flop \
	miff:- | composite -gravity southeast - $TEMPTHUMB $OUTPUTFILE
else
	cp $TEMPTHUMB $OUTPUTFILE
fi


rm $TEMPFILE1* $TEMPFILE2 $TEMPTHUMB

